### ISSUE TITLE
support for additional signers aws KMS or pkcs11

### ISSUE THREAD

**Issue by gbolo** [lestrrat-go/jwx#556]
Hi @lestrrat, Thanks for this amazing jw* library.

I see that you have a signer interface already:

```
// Signer generates the signature for a given payload.
type Signer interface {
	// Sign creates a signature for the given payload.
	// The scond argument is the key used for signing the payload, and is usually
	// the private key type associated with the signature method. For example,
	// for `jwa.RSXXX` and `jwa.PSXXX` types, you need to pass the
	// `*"crypto/rsa".PrivateKey` type.
	// Check the documentation for each signer for details
	Sign([]byte, interface{}) ([]byte, error)

	Algorithm() jwa.SignatureAlgorithm
}
```

with a few private implementations already (RSA, ECDSA, exc). Was wondering what your thoughts would be on including a more secure signer (that does not have direct access to the raw private key). Leveraging something like [AWS KMS](https://docs.aws.amazon.com/sdk-for-go/api/service/kms/#example_KMS_Sign_shared00), [Hashicorp Vault](https://www.vaultproject.io/api-docs/secret/transit#sign-data), or even [pkcs11](https://pkg.go.dev/crypto#Signer) (for HSM support) should accomplish this. Do you think it would be feasible to maintain one or more of these implementation(s)?

Thanks,
gbolo

**Comment by lestrrat** [comment]
Care to elaborate in pseudocode what this might look like? My hunch is that this is rather a sweeping change, and if it does get incorporated, it would have to be in a future major release.

As far as maintenance is concerned, it might be better for jwx to define an interface that these signers can use, and let interested personnel do the actual implementation?

**Comment by gbolo** [comment]
Hi @lestrrat,

After looking around a bit, you are right, this would be a rather sweeping change. Sorry to throw this on you ;)

My use case, is signing tokens (jwt). So currently I do something like:

```
// generate a raw key
rawECkey, _ := crypto.GenerateECKey("P-256")

// init a jwk object with it
key, _ := jwk.New(keyEC)

...
// use this key to sign a token
signedJwt, err := jwt.Sign(*token, signatureAlg, key)
```

Now the jwk interface is quite large, and it may not be possible to implement all
the functions in it. https://pkg.go.dev/github.com/lestrrat-go/jwx/jwk@v1.2.17#Key
Certainly the `Raw(interface{}) error` would be impossible.

Assuming we want to go this route, I would imagine it to look something like (for aws kms):

```
// init new aws session
sess, _ := session.NewSession(awsConfig)

// pass this session and some known key arn (or alias) to jwk
key, _ := jwk.NewAwsKmsKey(sess, "arn:aws:kms:us-west-2:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab")

// jwk.NewAwsKmsKey would use this info to init the kms service:  github.com/aws/aws-sdk-go/service/kms

...
// use this key to sign a token https://docs.aws.amazon.com/sdk-for-go/api/service/kms/#KMS.Sign
signedJwt, err := jwt.Sign(*token, signatureAlg, key)
```

please excuse the crudeness of my pseudocode ;)

If you can make this integration simpler somehow, then yes it would be great to leave it for the interested personnel to actually implement it. 

If this library would offer this integration already, it would certainly be a very unique feature. It may be of great interest to those operating in a very strict security landscape (such as government bodies, or financial institutions) that forbid applications from having access to the raw key.

**Comment by lestrrat** [comment]
Hmmmmmm. I'm not sure if cloaking KMS as a key would really solve this problem.
With my limited knowledge working on GCP KMS, I think it's not really representable as a key, but rather it's the whole package of jws+jwk (or jwe+jwk) combined.

So maybe it's really a new API that you want. Maybe it's something like a

```go
type Signer interface { ... } // nevermind that we already have an interface with the same name
type Verifier interface { ... }
kms := NewKMS(...)
kms.Sign(payload)
kms.Verify(payload)
```

And maybe this would have to be used in JWT as

```go
jwt.Parse(tokenPayload, jwt.WithVerifier(NewKMS(...))
jwt.Sign(token, jwt.WithSigner(NewKMS(...))
```

Either way, I think I'm going to have to obtain a sample KMS wrapper to think about how to fit it in the API. I'm ATM not really fluent in these tools, so if someone could come up with a reference implementation, that would be great.

**Comment by gbolo** [comment]
I think this new API you describe make sense. for a sample kms wrapper please see this project:

https://github.com/matelang/jwt-go-aws-kms

Once a new API is available to make this possible, I can try and come up with a reference implementation (time permitting).

Thanks for your input @lestrrat , you may close this issue if you like

**Comment by lestrrat** [comment]
I see. I think the current `jws.Signer` interface _almost_ works. we could cram it in to `jws.Sign()` if we allowed a calling convention like

```go
jws.Sign(payload, alg, nil, jws.WithSigner(kmsSigner))
```

albeit at some added complexity cost within the `jws.Sign()` method.

**Comment by lestrrat** [comment]
Hmmm, actually, if all payloads matching ES256 (just as an example) should use this new KMS, we can easily plug it using `jws.RegisterSigner()` and `jws.RegisterVerifier()`, I think. This is how we plug `es256k`

**Comment by gbolo** [comment]
> Hmmm, actually, if all payloads matching ES256 (just as an example) should use this new KMS, we can easily plug it using `jws.RegisterSigner()` and `jws.RegisterVerifier()`, I think. This is how we plug `es256k`

When a user creates an asymmetric KMS key, they choose the keytype from the following options:
`RSA_2048, RSA_3072, RSA_4096, ECC_NIST_P256, ECC_NIST_P384, ECC_NIST_P521, ECC_SECG_P256K1`. They also choose the key usage (sign, encrypt, decrypt, exc). The created key gets a unique ARN/ID (which is what is getting passed in)

The key type above can be retrieved from inspecting the public key (maybe other ways too):
```
pubOut, _ := kmsClient.GetPublicKey(context.Background(), &kms.GetPublicKeyInput{KeyId:&keyId})
fmt.Printf("Usage: %s\nSpec: %s\n", pubOut.KeyUsage, pubOut.KeySpec)
```

example output from an RSA key i created in KMS:
```
Usage: SIGN_VERIFY
Spec: RSA_2048
```

**Comment by lestrrat** [comment]
If I'm reading this right, I think RSA_2048 and the like are the key specs, whereas what we should be using when signing are the SigningAlgorithms such as kms.SigningAlgorithmSpecRsassaPkcs1V15Sha256 ? (not complaining or anything, I'm just making sure my understanding is correct)

**Comment by gbolo** [comment]
> 

yes that's correct, sorry I misunderstood your previous statement.

I will have to try to understand how to use the `jws.RegisterSigner()` that you mentioned. Thanks

**Comment by lestrrat** [comment]
@gbolo I think I can come up with a prototype soon, you can sit tight for a bit :)
You will have to do the testing through, as I don't really work with AWS and don't feel like setting up the environment ATM.

**Comment by lestrrat** [comment]
@gbolo Please checkout #557 along with https://github.com/lestrrat-go/jwx-aws-kms

I'm sure there are bugs, but the intended usage is in here: https://github.com/lestrrat-go/jwx-aws-kms/blob/main/jwxkms_test.go#L14

**Comment by gbolo** [comment]
Sounds good, I'll test it out ;) Thanks!

**Comment by gbolo** [comment]
@lestrrat thanks for quickly putting this together. I tried it but got this panic during sign:
```
panic: failed sign payload: failed to sign payload: failed to retrieve rsa.PrivateKey out of *jwxkms.RSA: expected rsa.PrivateKey or *rsa.PrivateKey, got *jwxkms.RSA [recovered]
	panic: failed sign payload: failed to sign payload: failed to retrieve rsa.PrivateKey out of *jwxkms.RSA: expected rsa.PrivateKey or *rsa.PrivateKey, got *jwxkms.RSA
```

this switch statement may be the problem here:

https://github.com/lestrrat-go/jwx/blob/fd774b7e524719c3ec686b7ffb8c780f078e3055/internal/keyconv/keyconv.go#L27


for reference, im using the same RSA 2048 KMS key; here is how it was called:
```
	sv := NewRSA(kms.NewFromConfig(awscfg)).
		WithAlgorithm(jwa.RS256).
		WithKeyID(kid)

	signed, err := jws.Sign(payload, jwa.RS256, sv.WithContext(ctx))
	if err != nil {
		panic(err.Error())
	}
```

I'm tired, will continue another day. Thanks again

**Comment by lestrrat** [comment]
That doesn't look right. Did you force Go to look at [jwx's gh-556 branch](https://github.com/lestrrat-go/jwx/pull/557)?

**Comment by lestrrat** [comment]
Huh, I haven't tried it myself, but there's an implementation of crypto.Signer using Google KMS.
If this works as advertised, you would be able to just plug it into jws.Sign without having to write any extra code.

https://github.com/salrashid123/signer

Just for the record

**Comment by lestrrat** [comment]
I just found this too https://github.com/lstoll/awskms

**Comment by lestrrat** [comment]
@gbolo Change of plans!

Please disregard #556, and see the following:

#559 fixes verification for `crypto.Signer`s. This was a bug
And I wrote a crypto.Signer wrapper for AWS KMS, which this time around I took the time to figure out how to test
https://github.com/jwx-go/crypto-signer/tree/main/aws

This is _much_ cleaner, and can be plugged in almost immediately.

**Comment by gbolo** [comment]
> @gbolo Change of plans!
> 
> Please disregard #557, and see the following:
> 
> #560 fixes verification for `crypto.Signer`s (this was a bug). And I wrote a crypto.Signer wrapper for AWS KMS, which this time around I took the time to figure out how to test https://github.com/jwx-go/crypto-signer/tree/main/aws. The one I wrote about earlier used the v1 API, this one uses v2.
> 
> This is _much_ cleaner, and can be plugged in almost immediately.

Just tested the signing, works perfectly! Thanks @lestrrat

**Comment by lestrrat** [comment]
Fixes to jwx has been merged via #560. Signers have been published to github.com/jwx-go/crypto-signers
